File: dotnet/Lexy.Poc/Lexy.Poc.Core/Parser/SourceCodeDocument.cs
using System;
using System.Linq;

namespace Lexy.Poc.Core.Parser
{
    public class SourceFile
    {
        public string FileName { get; }

        public SourceFile(string fileName)
        {
            FileName = fileName;
        }
    }

    public class SourceCodeDocument : ISourceCodeDocument
    {
        private int index;

        private Line[] code;

        public SourceFile File { get; private set; }

        public Line CurrentLine { get; private set; }

        public void SetCode(string[] lines, string fileName)
        {
            if (code != null)
            {
                throw new InvalidOperationException(
                    "Source code already set. Each LexyParser should only be used once. " +
                    "Scope should be managed by using ServiceProvider.OpenScope(). " +
                    "See SpecificationFileRunner.Create as example.");
            }

            index = -1;
            code = lines.Select((line, index) => new Line(index, line)).ToArray();
            File = new SourceFile(fileName);
        }

        public bool HasMoreLines()
        {
            return index < code.Length - 1;
        }

        public Line NextLine()
        {
            if (index >= code.Length)
            {
                throw new InvalidOperationException("No more lines");
            }

            CurrentLine = code[++index];
            return CurrentLine;
        }

        public void Reset()
        {
            CurrentLine = null;
        }
    }
}

File: typescipt/src/parser/sourceCodeDocument.ts


export class SourceCodeDocument extends ISourceCodeDocument {
   private Line[] code;
   private number index;
   private SourceFile file;

   public Line CurrentLine { get; private set; }

   public setCode(lines: string[], fileName: string): void {
     index = -1;
     file = new SourceFile(fileName);
     code = lines.Select((line, index) => new Line(index, line, file)).ToArray();
   }

   public hasMoreLines(): boolean {
     return index < code.Length - 1;
   }

   public nextLine(): Line {
     if (index >= code.Length) throw new Error(`No more lines`);

     CurrentLine = code[++index];
     return CurrentLine;
   }

   public reset(): void {
     CurrentLine = null;
   }
}
